Hall (New Talk): the PR

Fang- opened the pull request:
Nearly eight months later, it's finally here!

`talk` as we know and love it has been split into `hall` (the backend) and `talk` (the cli frontend). Structures and interfaces got overhauled, library and marks redone, and dependent apps have been updated to match. Fora now has its own super simple app, and uses `hall` to re-publish the content to its own special-purpose circles.

Contains updated JS files for webtalk. PRs for webtalk and fora are on their way.

There's still a lot of things I'd like to change/improve upon, but I'm confident enough in the current state of things that I want to try getting this into the breach. No doubt the chaos of actual usage will reveal some more kinks, but those can be easily patched over the air. I'm looking forward to finally getting this into the hands of the people.

(I noticed just now that some files also have changes to remove trailing whitespace. Didn't intend for that to be included in my commits here, but considering this currently merges cleanly into master, that probably isn't too big a deal.)

Depends on #451.  
Contains (temporary?) workarounds for #447 (investigation ongoing).

Questions and concerns very much welcome.

ohAitch commented:
Just to double check,

<details><summary>"doing  =,  talk  on the top-level of this /lib file causes a bail: fail (^:"</summary>

```
~palnup_livpyl= i'm still getting a weird find.$.+2 on a simple json reparser
~palnup_livpyl= like this:
~palnup_livpyl= (of lin+(ot pat+bo txt+so ~) ~)
~palnup_livpyl= it points to (ot ...) specifically, but i see nothing wrong with
-------[61.160]
~palnup_livpyl= it
~pittyp-datfyn# (of lin+(ot pat+bo txt+so ~) ~):jo__<1.lrf {jon/#1 <1.fhl {wer
~palnup_livpyl= yeah it compiles just fine out of context, but there's no part
~palnup_livpyl= of the context i can find that's messing this up
~palnup_livpyl= nevermind, i lied
-------[61.165]
~pittyp-datfyn= huh, are  of ot bo so  all the right things?
~palnup_livpyl= i was doing =,  dejson-soft  =,  talk
~palnup_livpyl= doing that the other way around fixed it
~palnup_livpyl= even though there's nothing in /sur/talk that overwrites any of
~palnup_livpyl= those?
-------[61.170]
~palnup_livpyl= (this is also after putting those in arms because doing =,  talk
~palnup_livpyl= on the top-level of this /lib file causes a bail: fail (^: )
~pittyp-datfyn= uh
~pittyp-datfyn really needs to fix lib/ sur/ handling
~palnup_livpyl= yeah i figured it was some in-dev weirdness
-------[61.175]
~palnup_livpyl= hm, getting the same bail: fail now whenever i try to use a mark
~palnup_livpyl= that calls the lib from earlier
~palnup_livpyl= i guess this means they both compile, hurrah, but pff
~palnup_livpyl= removing the stray =, talk at the top of the mark doesn't
~palnup_livpyl= resolve this either
-------[61.180]
~palnup_livpyl= "port to remaint" they said, "it'll be cool" they said
```

</details>

from the breach checklist to-file section got resolved yes?

cgyarvin commented:
If you have to ask, just try it :-)

Sent from my iPhone

> On Nov 9, 2017, at 5:26 PM, Anton Dyudin <notifications@github.com> wrote:
> 
> Just to double check,
> 
> "doing =, talk on the top-level of this /lib file causes a bail: fail (^:"
> from the breach checklist to-file section got resolved yes?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

Fang- commented:
I can put the `=,` on the top level in `/lib/hall-json` just fine now, but the marks still suffer from problems when using `=,`.

For example, changing [this](https://github.com/Fang-/arvo/blob/remaint-merge-talk/mar/hall/prize.hoon#L13) to use `=,` still results in a `bail: fail` as was previously discussed.

belisarius222 commented:
@Fang- there are some conflicts on this PR. I'm about to test it out using the urbit/fleet repo.

Fang- commented:
Fixed a small regression and tested + improved client behavior on moons. Also removed the lib file's public domain notice, since I expect that to have to go.

As far as I'm concerned, this is once again ready to be merged. @belisarius222 may want to test moons himself to confirm my fixes.

Consider this a request for comments. If anyone's opposed to any part of this making it into the next breach, or feels like something crucial is still missing, please speak up now. (:

ohAitch commented on lib/hall-json.hoon:
```diff
@@ -0,0 +1,605 @@
+    ==
+  ::
+  ++  sorc                                              ::>  source
+    ;~(plug circ rang)
```
```
++  pont
  ;~  pfix  fas
    %+  sear
      |=  a=coin  ^-  (unit place)
      ?+(a ~ [%$ %da @da] `p.a, [%$ %ud @ud] `p.a)
    nuck:so
  ==
++  sorc  ;~(plug circ (punt ;~(plug pont (punt pont))))
```

Fang- replied:
That's much more concise, thanks!

Fang- commented on sur/hall.hoon:
```diff
@@ -0,0 +1,259 @@
+::+|
+::
+::TODO  rename
+++  naem  term                                          ::<  circle name
```
Feels pretty ambiguous with `circle` itself imo. "A `circle` is a `ship` and a `circ`" is weird.

Not even sure if I still want to see this renamed...

ohAitch replied:
I think the default for not finding a good name should be leaving it as `term`/`@tas`, not introducing phonetic ambiguity. Is just `name` even used anywhere?

Fang- replied:
It is, in zuse. When I wrote this, that wasn't namespaced yet. I guess I have no reason not to rename this to `++name` now that it is.

ohAitch replied:
ah, the ames one. I'd still push for `chan` or sth that points that this is a chat room name, but either way definitely not `naem`

Fang- commented on app/talk.hoon:
```diff
@@ -1,272 +1,654 @@
+::+|
+::
+++  our
+  (true-self our.bol)
```
Hmm, I guess that's fair. This was intended as "if you're using `our`, you want to use your planet", but I agree using a 3lw may still be weird. `self` has some minor semantic clashing (aka `this`), but is probably fine here.

cgyarvin commented:
Yeah naem looks like something I’d do as a debugging hack, then clear up later...

Sent from my iPhone

> On Dec 12, 2017, at 2:01 PM, Anton Dyudin <notifications@github.com> wrote:
> 
> @ohAitch commented on this pull request.
> 
> In sur/hall.hoon:
> 
> > +::::  /sur/hall/hoon
> +  ::
> +|%
> +::
> +::TODO  use different words for different kinds of burdens
> +::TODO  rename det/delta in most place? they may be (different kinds of) deltas,
> +::      but location in control flow already indicates delta-ness.
> +::
> +::>  ||
> +::>  ||  %wrappers
> +::>  ||
> +::>    wrapper molds, for semantic clarity.
> +::+|
> +::
> +::TODO  rename
> +++  naem  term                                          ::<  circle name
> ah, the ames one. I'd still push for chan or sth that points that this is a chat room name, but either way definitely not naem
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
